pkg/router/general/search.go

```diff
@@ -0,0 +1,288 @@
+package general
+
+import (
+	"errors"
+	"math/big"
+	"net/http"
+
+	"github.com/jackc/pgx/v5/pgtype"
+	"github.com/jykuo-love-shiritori/twp/db"
+	"github.com/jykuo-love-shiritori/twp/minio"
+	"github.com/jykuo-love-shiritori/twp/pkg/common"
+	"github.com/labstack/echo/v4"
+	"go.uber.org/zap"
+)
+
+type searchParams struct {
+	Q          string `query:"q"`
+	MinPrice   int32  `query:"minPrice"`
+	MaxPrice   int32  `query:"maxPrice"`
+	HaveCoupon bool   `query:"haveCoupon"`
+	SortBy     string `query:"sortBy"`
+	Order      string `query:"order"`
+	common.QueryParams
+}
+
+func (q *searchParams) Validate() error {
+	if q.Q == "" {
+		return errors.New("search query is empty")
+	}
+	if q.MinPrice < 0 || q.MaxPrice < 0 || q.MinPrice > q.MaxPrice {
+		return errors.New("invalid price range")
+	}
```

Comment:
do user must set min_price and max_price?

Reply:
since it default zero, this only check negative and min>max